### PR TITLE
Refactor FMU: move mixer handling into a library

### PR DESCRIPTION
**Pull request by bkueng**
Currently all actuator output drivers handle mixer loading, mixer parameters and actuator controls subscriptions with everything around that themselves, leading to duplicated code with minor differences.
This PR is a step towards improving that by pulling these parts out of fmu into a library that can be used by other modules.
It is a pure refactoring, except for:
- removal of `PX4FMU::write` (https://github.com/PX4/Firmware/commit/9c842514952c84b33f24dac43ce478e6788c1464). I don't see it being used anywhere.
- mixer loading is now thread-safe (https://github.com/PX4/Firmware/commit/418021b0af4c05bbdd39b241bf055b485a57b9a8)

### Testing
I tested:
- normal pwm on Pixhawk 4, including safety button
- Pixracer, including an ESC calibration
- oneshot on omnibus
- mixer param changes

@Igor-Misic this serves as preparation for DShot.
I tend towards adding it as a new module (as opposed to integrate into fmu), as the fmu driver contains various things we won't need. On the other hand however, in the current form we would lose input capture pin configuration ~~and the safety button~~.

@dagar one problem I found with the actuator_controls uORB scheduling is that you can end up your module not being scheduled at all. For example with an RC passthrough mixer and no RC connected. If the module wants to do other tasks as well we need to ensure it's run periodically even w/o uORB updates.

A careful review is appreciated.

### CODE REVIEW

**Comment by dagar** [comment]
> @dagar one problem I found with the actuator_controls uORB scheduling is that you can end up your module not being scheduled at all. For example with an RC passthrough mixer and no RC connected. If the module wants to do other tasks as well we need to ensure it's run periodically even w/o uORB updates.

We should get those other tasks removed from fmu. https://github.com/PX4/Firmware/pull/10119

File: src/lib/mixer_module/mixer_module.hpp [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+	const bool _support_esc_calibration;
+
+	bool _wq_switched{false};
+
+	OutputModuleInterface &_interface;
+
+	perf_counter_t _control_latency_perf;
+
+	DEFINE_PARAMETERS(
+		(ParamInt<px4::params::MC_AIRMODE>) _param_mc_airmode,   ///< multicopter air-mode
```
Comment (dagar) [inline src/lib/mixer_module/mixer_module.hpp]: We should think of a different way to handle this eventually. Seeing the "Multicopter Attitude Control" param group on a plane deeply bothers me.

File: src/lib/mixer_module/mixer_module.cpp [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+	}
+
+	/* do mixing */
+	float outputs[MAX_ACTUATORS] {};
+	const unsigned mixed_num_outputs = _mixers->mix(outputs, MAX_ACTUATORS);
+
+	/* the PWM limit call takes care of out of band errors, NaN and constrains */
+	uint16_t output_limited[MAX_ACTUATORS] {};
+
+	pwm_limit_calc(_throttle_armed, armNoThrottle(), mixed_num_outputs, _reverse_output_mask,
```
Comment (dagar) [inline src/lib/mixer_module/mixer_module.cpp]: Is there anything PWM specific left here? pwm_limit_calc -> limit_calc

**Comment by dagar** [comment]
How do you see this working in cases like uavcan? I'm wondering if having the uORB subscription callbacks in mixer_module is going to be too restrictive for things like uavcan.

File: src/lib/mixer_module/mixer_module.hpp (hunk 1 above)
Comment (bkueng) [reply src/lib/mixer_module/mixer_module.hpp]: Agreed. Maybe just move the param to a mixer group?

File: src/lib/mixer_module/mixer_module.cpp (hunk 2 above)
Comment (bkueng) [reply src/lib/mixer_module/mixer_module.cpp]: I don't see anything. The diff will be quite large though, but I can do it.

**Comment by bkueng** [comment]
> How do you see this working in cases like uavcan? I'm wondering if having the uORB subscription callbacks in mixer_module is going to be too restrictive for things like uavcan.

So far I see these options:
- poll only on actuator groups 0 (or 1), and if not used/published run at a fixed rate
- moving the subscription callbacks out of the library back into the module, but that again leads to more duplication.
- Independently schedule MixingOutput by making it a separate work item. But this leads to more overhead and complexity (switching work queues for example), so I tend towards the first option.

**Comment by jlecoeur** [comment]
I think this goes in the right direction, although I have not reviewed carefully. 
I would tend towards the second option (i.e. to not impose callback or scheduling in the library). And later, remove the duplicated uses of this library by centralizing mixing in one module running on FMU.

File: src/lib/mixer_module/mixer_module.cpp (hunk 2 above)
Comment (bkueng) [reply src/lib/mixer_module/mixer_module.cpp]: Done.

**Comment by bkueng** [comment]
Rebased.

> I would tend towards the second option (i.e. to not impose callback or scheduling in the library). And later, remove the duplicated uses of this library by centralizing mixing in one module running on FMU.

I made it configurable now so the user of the library can decide.

**Comment by jorge789** [comment]
**Tested on NXP FMUK66 V3**
**Modes Tested**

Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL: Good.

**- Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoin activate RTL and see landing behaviour.

**Notes:**
No issues noted, good flight in general.

**Log:**
https://review.px4.io/plot_app?log=63165b52-49d0-45af-9c80-451bdc0374cd

**Comment by jorge789** [comment]
### Tested on nano V5:
**Modes Tested**

Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL: Good.

**- Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoin activate RTL and see landing behaviour.

**Notes:**
The height of the vehicle and what QGC said did not coincide. In the mission the height was 35 meters and the height of the actual vehicle was approximately 25 meters. The error was only looked at Altitude, stabilized and mission,in position the error could not be seen.

**Log:**
https://review.px4.io/plot_app?log=6e6f1be3-5f78-4e7e-9281-777aa912329c

**Pull request merged**